File: bench/microbench3.java
import java.util.Date;
import java.lang.Math;


class Dim {
	int x;
	int y;
	int z;
	void f() {
	}
}

class microbench3 {
	static int global_x = 0;
	static int global_y = 0;
	static int global_z = 0;

	static int testSimpleLoop() {
		int i;
		for (i = 0; i < 100000000; i++) {
		}
		return i;
	}
	
	static int testLocalVariable() {
		int y = 0;
		int i;
		for (i = 0; i < 100000000; i++) {
			y = i;
		}
		return y;
	}

	static int testGlobalVariable() {
		int i;
		for (i = 0; i < 100000000; i++) {
			global_y = i;
		}
		return global_y;
	}

	static String testStringAssignment() {
		String s = "";
		int i;
		for (i = 0; i < 100000000; i++) {
			s = "A";
		}
		return s;
	}

	static int testIntegerOperation() {
		int y = 0;
		int i;
		for (i = 0; i < 100000000; i++) {
			y += 1;
		}
		return y;
	}

	static double testFloatOperation() {
		double f = 0.0;
		int i;
		for (i = 0; i < 100000000; i++) {
			f += 0.1;
		}
		return f;
	}

	static void func0() {}

	static int testFunctionCall() {
		int i = 0;
		for (i = 0; i < 100000000; i++) {
			func0();
		}
		return i;
	}

	static int func1() { return 1; }

	static int testFunctionReturn() {
		int res = 0;
		int i;
		for (i = 0; i < 100000000; i++) {
			res = func1();
		}
		return res;
	}

	static int testMathFabs() {
		int i;
		for (i = 0; i < 100000000; i++) {
			Math.abs(-1.0);
		}
		return i;
	}

	/*
	static int testCallFunctionObject() {
		int i;
		for(i = 0; i< 100000000; i++) {
			func0();
		}
		return i;
	}
	*/

	static int fibo(int n) {
		if (n < 3) return 1;
		return fibo(n - 1) + fibo(n - 2);
	}
	
	static int testObjectCreation() {
		int i;
		for (i = 0; i < 100000000; i++) {
			Dim d = new Dim();
		}
		return i;
	}

	static Dim testFieldVariable() {
		Dim d = new Dim();
		int i;
		for (i = 0; i < 100000000; i++) {
			d.y = 1;
		}
		return d;
	}

	static int testMethodCall() {
		int i;
		Dim d = new Dim();
		for (i = 0; i < 100000000; i++) {
			d.f();
		}
		return i;
	}

	static long mops(long score) {
		if (score < 1) return 0;
		return (10000000 / score) / 100;
	}
	
	public static void main(String args[]) {
		long s = System.currentTimeMillis();
		testSimpleLoop();
		long t = System.currentTimeMillis() - s;
		long score = t;
		System.out.println("SimpleLoop[mops, ms] : " + mops(score) + ", t = " + t);
		long delta = t;
		//
		s = System.currentTimeMillis();
		testLocalVariable();
		t = System.currentTimeMillis() - s;
		score = t - delta;
		System.out.println("LocalVariable[mops, ms] : " + mops(score) + ", t = "  + t);
		//
		s = System.currentTimeMillis();
		testGlobalVariable();
		t = System.currentTimeMillis() - s;
		score = t - delta;
		System.out.println("GlobalVariable[mops, ms] : " + mops(score) + ", t = "  + t);
		//
		s = System.currentTimeMillis();
		testStringAssignment();
		t = System.currentTimeMillis() - s;
		score = t - delta;
		System.out.println("StringAssignment[mops, ms] : " + mops(score) + ", t = "  + t);
		//
		s = System.currentTimeMillis();
		testIntegerOperation();
		t = System.currentTimeMillis() -s ;
		score = t - delta;
		System.out.println("IntegerOperation[mops, ms] : " + mops(score) + ", t = "  + t);
		//
		s = System.currentTimeMillis();
		testFloatOperation();
		t = System.currentTimeMillis() -s ;
		score = t - delta;
		System.out.println("FloatOperation[mops, ms] : " + mops(score) + ", t = "  + t);
		// 
		s = System.currentTimeMillis();
		testFunctionCall();
		t = System.currentTimeMillis() - s;
		score = t - delta;
		System.out.println("FunctionCall[mops, ms] : " + mops(score) + ", t = "  + t);
		//
		s = System.currentTimeMillis();
		testFunctionReturn();
		t = System.currentTimeMillis() - s;
		score = t - delta;
		System.out.println("FunctionReturn[mops, ms] : " + mops(score) + ", t = "  + t);
		//
		s = System.currentTimeMillis();
		testMathFabs();
		t = System.currentTimeMillis() - s;
		score = t - delta;
		System.out.println("NativeFunction[mops, ms] : " + mops(score) + ", t = "  + t);
		//
		/*
		s = System.currentTimeMillis();
		testCallFunctionObject();
		t = System.currentTimeMillis() - s;
		score = t - delta;
		System.out.println("CallFunctionObject[mops, ms] : " + mops(score) + ", t = "  + t);
		*/
		//
		System.out.println("Object-Oriented Programming");
		s = System.currentTimeMillis();
		testObjectCreation();
		t = System.currentTimeMillis() -s ;
		score = t - delta;
		System.out.println("ObjectCreation[mops, ms] : " + mops(score) + ", t = "  + t);
		//
		s = System.currentTimeMillis();
		testFieldVariable();
		t = System.currentTimeMillis() -s ;
		score = t - delta;
		System.out.println("FieldVariable[mops, ms] : " + mops(score) + ", t = "  + t);
		//
		s = System.currentTimeMillis();
		testMethodCall();
		t = System.currentTimeMillis() -s ;
		score = t - delta;
		System.out.println("MethodCall[mops, ms] : " + mops(score) + ", t = "  + t);
		//
		System.out.println("MISC (this is not microbench)");
		s = System.currentTimeMillis();
		fibo(36);
		t = System.currentTimeMillis() - s;
		score = t;
		System.out.println("fibonacci(36)[mops, ms] : " + mops(score) + ", t = "  + t);
	}
}

File: bench/microbench3.lua
-- Lua microbench by kimio kuramitsu

function testSimpleLoop()
	for i = 0, 100000000 do
	end
	return i
end

function testLocalVariable()
	y = 0
	for i = 0, 100000000 do
		y = i
	end
	return y
end

global_x = 0
global_y = 0
global_z = 0

function testGlobalVariable()
	for i = 0, 100000000 do
		global_y = 1
	end
	return global_y
end

function testStringAssignment()
	str = "";
	for i = 0, 100000000 do
		str = "A";
	end
	return str;
end

function testIntegerOperation()
	y = 0;
	for i = 0, 100000000 do
		y = y + 1;
	end
	return y;
end

function testFloatOperation()
	f = 0.0;
	for i = 0, 100000000 do
		f = f + 0.1;
	end
	return f;
end


function func0()
end

function testFunctionCall()
	i = 0
	for i = 0, 100000000 do
		func0()
	end
	return i
end

function func1()
	return 1
end

function testFunctionReturn()
	res = 0
	for i = 0, 100000000 do
		res = func1()
	end
	return res
end

function testMathFabs()
	for i = 0, 100000000 do
		math.abs(-0.1);
	end
	return i
end

function testCallFunctionObject()
	f = func1;
	for i = 0, 100000000 do
		f();
	end
	return i;
end

Dim = {}
function Dim.new()
	local instance = {
		x = 0,
		y = 0,
		z = 0
	}
	return instance
end

function Dim:f()
end

--

function testObjectCreation()
	for i = 0, 100000000 do
		d = Dim.new()
	end
	return d
end

function testFieldVariable()
	d = Dim.new()
	for i = 0, 100000000 do
		d.y = d.y + 1
	end
	return d
end


function testMethodCall()
	d = Dim.new()
	for i = 0, 100000000 do
		d.f();
	end
	return d
end


function fibo(n)
	if n < 3 then
		return 1
	else
		return fibo(n-1) + fibo(n-2)
	end
end

function mops(score) 
	--if n < 1.0 then 
	--	return 0.0;
	--end
	return (10000000 / score) / 100.0;
end

function run()
	s = os.clock() * 1000
	testSimpleLoop()
	t = os.clock() * 1000 - s
	score = t
	print("SimpleLoop[mops,ms]: ", string.format("%f", mops(score)), string.format("%d", t))
	delta = t
	--
	s = os.clock() * 1000
	testLocalVariable()
	t = os.clock() * 1000 - s
	score = t - delta
	print("LocalVariable[mops,ms]:", string.format("%f", mops(score)), string.format("%d", t))
	--
	s = os.clock() * 1000
	testGlobalVariable()
	t = os.clock() * 1000 - s
	score = t - delta
	print("GlobalVariable[mops,ms]:", string.format("%f", mops(score)), string.format("%d", t))
	--
	s = os.clock() * 1000
	testStringAssignment()
	t = os.clock() * 1000 - s
	score = t - delta
	print("StringAssignment[mops,ms]:", string.format("%f", mops(score)), string.format("%d", t))
	--
	s = os.clock() * 1000
	testIntegerOperation()
	t = os.clock() * 1000 - s
	score = t - delta
	print("IntegerOperation[mops,ms]:", string.format("%f", mops(score)), string.format("%d", t))
	--
	s = os.clock() * 1000
	testFloatOperation()
	t = os.clock() * 1000 - s
	score = t - delta
	print("FloatOperation[mops,ms]:", string.format("%f", mops(score)), string.format("%d", t))
	--
	s = os.clock() * 1000
	testFunctionCall()
	t = os.clock() * 1000 - s
	score = t - delta
	print("FunctionCall[mops,ms]:", string.format("%f", mops(score)), string.format("%d", t))
	--
	s = os.clock() * 1000
	testFunctionReturn()
	t = os.clock() * 1000 - s
	score = t - delta
	print("FunctionReturn[mops,ms]:", string.format("%f", mops(score)), string.format("%d", t))
	--
	s = os.clock() * 1000
	testCallFunctionObject()
	t = os.clock() * 1000 - s
	score = t - delta
	print("CallFunctionObject[mops,ms]:", string.format("%f", mops(score)), string.format("%d", t))
	--
	s = os.clock() * 1000
	testMathFabs()
	t = os.clock() * 1000 - s
	score = t - delta
	print("NativeFunction[mops,ms]:", string.format("%f", mops(score)), string.format("%d", t))
	--
	print("Object-Oriented Programming")
	s = os.clock() * 1000
	testObjectCreation()
	t = os.clock() * 1000 - s
	score = t - delta
	print("ObjectCreation[mops,ms]:", string.format("%f", mops(score)), string.format("%d", t))
	s = os.clock() * 1000
	testFieldVariable()
	t = os.clock() * 1000 - s
	score = t - delta
	print("FieldVariable[mops,ms]:", string.format("%f", mops(score)), string.format("%d", t))

	s = os.clock() * 1000
	testMethodCall()
	t = os.clock() * 1000 - s
	score = t - delta
	print("MethodCall[mops,ms]:", string.format("%f", mops(score)), string.format("%d", t))

	--
	print "MISC (this is not microbench)";
	s = os.clock() * 1000
	fibo(36)
	t = os.clock() * 1000 - s
	score = t - delta
	print("fibonacci[mops,ms]:", string.format("%f", mops(score)), string.format("%d", t))
end

run()
